pkg/segment/query/iqr/iqrreader.go

```diff
@@ -0,0 +1,412 @@
+		if reflect.DeepEqual(iqrRdr, otherRdr) {
+			return nil
+		}
+
+		resultRdr := NewMultiIQRReader()
+
+		err := resultRdr.AddSingleIQRReader(iqr)
+		if err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add iqr to resultRdr; err=%v", err)
+		}
+
+		err = resultRdr.AddSingleIQRReader(otherIQR)
+		if err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add otherIQR to resultRdr; err=%v", err)
+		}
+
+		iqr.reader = resultRdr
+
+		return nil
+	}
+
+	if iqrRdr.IsSingleReader() {
+		iqrRdr.readerMode = ReaderModeMultiReader
+		if err := iqrRdr.AddSingleIQRReader(otherIQR); err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add otherIQR to iqr; err=%v", err)
+		}
+	}
+
+	if otherRdr.IsSingleReader() {
+		if err := iqrRdr.AddSingleIQRReader(otherIQR); err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add otherIQR to iqr; err=%v", err)
+		}
+	} else {
+		for readerId, reader := range otherRdr.readerIdToReader {
+			if err := iqrRdr.AddReader(readerId, reader); err != nil {
+				return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add reader to iqr; err=%v", err)
+			}
+		}
+
+		for segEnc, readerId := range otherRdr.encodingToReaderId {
+			if err := iqrRdr.AddSegEncodingToReader(segEnc, readerId); err != nil {
+				return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add seg encoding to reader; err=%v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (iqrRdr *IQRReader) validate() error {
+	if iqrRdr.IsSingleReader() {
+		if iqrRdr.reader == nil {
+			return fmt.Errorf("iqrReader.validate: reader is nil in single reader mode")
+		}
+		return nil
+	}
+
+	for segEnc, readerId := range iqrRdr.encodingToReaderId {
+		if _, ok := iqrRdr.readerIdToReader[readerId]; !ok {
+			return fmt.Errorf("iqrReader.validate: readerId=%v not found for segEnc=%v", readerId, segEnc)
+		}
+	}
+
+	return nil
+}
+
+func (iqrRdr *IQRReader) readColumnsForRRCs(segKey string, vTable string, rrcs []*utils.RecordResultContainer,
+	qid uint64, ignoredCols map[string]struct{}, columnName *string) (map[string][]utils.CValueEnclosure, error) {
+	if len(rrcs) == 0 {
+		return nil, nil
+	}
+
+	if iqrRdr.IsSingleReader() {
+		return nil, fmt.Errorf("iqrReader.ReadAllColsForRRCs: single reader mode not supported")
+	}
+
+	rrcReaderGroups := make(map[utils.T_SegReaderId][]*rrcWithIndex)
+
+	for i, rrc := range rrcs {
+		readerId := iqrRdr.encodingToReaderId[rrc.SegKeyInfo.SegKeyEnc]
+		_, ok := rrcReaderGroups[readerId]
+		if !ok {
+			rrcReaderGroups[readerId] = make([]*rrcWithIndex, 0, len(rrcs))
+		}
+		rrcReaderGroups[readerId] = append(rrcReaderGroups[readerId], &rrcWithIndex{rrc: rrc, index: i})
+	}
+
+	resultChan := make(chan readerResult, len(rrcReaderGroups))
+
+	wg := sync.WaitGroup{}
+
+	for readerId, rrcsWithIndex := range rrcReaderGroups {
+		reader := iqrRdr.readerIdToReader[readerId]
+
+		wg.Add(1)
+		go func(readerId utils.T_SegReaderId, rrcsWithIndex []*rrcWithIndex, reader record.RRCsReaderI) {
```

Comment:
I was thinking there'd only be one reader per segkey. Is that incorrect? The method you're in takes a segkey and a list of RRCs, so all the RRCs should be for the same segkey, right?

If there will only be one reader with any `rrcsWithIndex`, then using goroutines and a waitgroup seems unnecessary

Reply:
Ohh yes, you are right, for some reason, I was under the impression that the IQR can have RRCs from multiple readers (which is true) and is also same for this function (RRCs contain multiple readers). But since here we are filtering by `SegKey`, yes we don't need that code. I will refactor